lxd/db/node.go

```diff
@@ -1139,6 +1139,10 @@ func (c *ClusterTx) GetNodeWithLeastInstances(ctx context.Context, members []Nod
 		}
 	}
 
+	if member == nil {
```

Comment:
We also need to update callers of `evacuateClusterSelectTarget` which will now get an error when no member was selected, so for example in `evacuateInstances` we have:

```go
		targetMemberInfo, err := evacuateClusterSelectTarget(ctx, opts.s, opts.gateway, inst, candidateMembers)
		if err != nil {
			return err
		}

		// Skip migration if no target available.
		if targetMemberInfo == nil {
			l.Warn("No migration target available for instance")
			continue
		}
```

Previously we would have continued here, but now it will fail with an error so need to change it to:

```go
		targetMemberInfo, err := evacuateClusterSelectTarget(ctx, opts.s, opts.gateway, inst, candidateMembers)
		if err != nil {
	        if api.StatusErrorCheck(err, http.StatusNotFound) {
                // Skip migration if no target available.
				l.Warn("No migration target available for instance")
			    continue
			}

			return err
		}
```

Reply:
I am struggling with if it is an `http.StatusNotFound` or `http.BadRequest`. 

In case of the instance movement, the user would request `lxc move c1 --target foo`. When getting back a 404 from this operation I would think either `c1` or `foo` doesn't exist. @tomponline what do you think?

Reply:
You could push that detection into `evacuateClusterSelectTarget` if you prefer and then continue to return nil here.

Reply:
It should remain http.StatusNotFound in the context of the candidate search though.

Reply:
the only caller of `evacuateClusterSelectTarget` is `evacuateInstances` and it already has a check it not member could be found:

```go
// Skip migration if no target available.
if targetMemberInfo == nil {
	l.Warn("No migration target available for instance")
	continue
}
```

I would then again return nil from `GetNodeWithLeastInstances` and explicitly return a 404 for both instance and instances endpoint in case there isn't any node available.

Reply:
So `GetNodeWithLeastInstances` needs to remain as it is in this PR, returning a not found error.

This will change the return of `evacuateClusterSelectTarget` as it will now return an error rather than nil in that case.

So I propose adding the not found detection to `evacuateClusterSelectTarget` so that it continues to return nil in that case so that the caller of `evacuateClusterSelectTarget` doesn't need to change.

Reply:
Seems more reasonable, otherwise the error message in both endpoints needs to be duplicated.

Reply:
Yes, that was what we were trying to avoid originally.

Reply:
Changes applied.